### PR TITLE
Options panel and app bar menu

### PR DESCRIPTION
**Pull request by TomTirapani**

### CODE REVIEW

**Comment by TomTirapani** [comment]
This branch has been updated to work with (and simplified by) the latest Forms model

File: cmp/form/validation/Rule.js [hunk 1]
```diff
@@ -59,6 +59,7 @@ export class Rule {
     // Implementation
     //------------------------------
     isActive(field) {
+        if (!field.formModel) return;
```
Comment (lbwexler) [inline cmp/form/validation/Rule.js]: This should explicitly return false

File: core/XH.js [hunk 2]
```diff
@@ -327,6 +327,11 @@ class XHClass {
     //---------------------------
     // Miscellaneous
     //---------------------------
+    /** Show a dialog to for users to set app options. */
```
Comment (lbwexler) [inline core/XH.js]: comment typo

File: core/appcontainer/OptionsDialogModel.js [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    @computed
+    get hasOptions() {
+        return !!this.options.length;
+    }
+
+    //-------------------
+    // Change management
+    //-------------------
+    @action
+    refreshChangedFields() {
```
Comment (lbwexler) [inline core/appcontainer/OptionsDialogModel.js]: I don't get the need for the extra state in the changes array.  Couldn't we just walk through the fields, both for the computed "requiresRefresh" and saveAsync()?

File: desktop/cmp/appbar/AppBar.js [hunk 4]
```diff
@@ -47,6 +47,9 @@ export class AppBar extends Component {
          */
         hideAdminButton: PT.bool,
```
Comment (lbwexler) [inline desktop/cmp/appbar/AppBar.js]: Can we take advantage of this change to combine the admin button/feedback button/theme button/options button within a single framework menu of a single button?  Think its too many framework buttons that apps will rarely want to dedicate their prime real estate to.

**Comment by lbwexler** [comment]
Think this is going to give us some great consistency, and be a huge improvement for apps.
Some comments inline: 

+ Do getValue() and setValue() need to be getValueAsync() and setValueAsync()?

**Comment by lbwexler** [comment]
I am ok with this change and would like to get it in.  It actually has a couple of form fixes that are important to take. @amcclain -- thoughts?

My one concern is that I would really like to combine all of the app bar buttons in this version:
        hideAdminButton: PT.bool,
        hideOptionsButton: PT.bool,
        hideFeedbackButton: PT.bool,
       hideThemeButton: PT.bool
       hideThemeButton: PT.bool

into a single "Hoist Menu", equivalent to Chromes three button menu.  

It can include menus items for all of these, and we can generally be more consistent across desktop and mobile, where we definitely don't have room for all these buttons. 

Note:   *Apps that wish to add individual buttons can still do so -- they are all part of the public API*

**Comment by TomTirapani** [comment]
@lbwexler I just updated this branch to move the various app bar buttons into a 'App Menu'.

**Comment by lbwexler** [comment]
Made some additional changes to harmonize with how we are specifying forms in RestGrid

Removing WIP as we are getting ready for merge

**Comment by lbwexler** [comment]
I think the options panel itself is great and ready to be merged now

Want to get @atm thoughts on moving all the toolbar items into a single menu.  

Also wonder if there is an easier way to mix and match the items and also combine them with application logic?  Maybe appModel should have a method 'getAppMenuItems(){}, which can be specified to make the hamburger menu?  And a set of constants somewhere (on the appContainer?) which define the available built-in partials?

**Comment by amcclain** [comment]
I will take a look.  Mainly would really like to favor consistency so key actions (especially launch admin) are readily accessible across apps in the same place.  We tried the hamburger menu recently in a desktop app and I was surprised at how much it threw off my muscle memory and expectations vs. all other apps.  

In general, I think it's a good differentiation between desktop and mobile that desktop gives you quicker access to more functions since you have so much more space available.

But I will look of course.  I was thinking that one advantage of the options panel is that the theme switching had a more natural place to live, so I was hoping that would effectively be replaced by the options button.

**Comment by lbwexler** [comment]
Got it -- We can certainly roll back the AppBar changes for future consideration, and simply add a new button for this to AppBar.  Just want to get it merged....

**Comment by amcclain** [comment]
> I will take a look. Mainly would really like to favor consistency so key actions (especially launch admin) are readily accessible across apps in the same place. We tried the hamburger menu recently in a desktop app and I was surprised at how much it threw off my muscle memory and expectations vs. all other apps.

and for the record I decided to change my mind about this 😛

**Pull request merged**